src/Storages/System/StorageSystemZooKeeper.cpp

```
@@ -336,6 +349,50 @@ static bool isPathNode(const ActionsDAG::Node * node)
     return node->result_name == "path";
 }
 
+static void extractNameImpl(const ActionsDAG::Node & node, String & res, ContextPtr context)
+{
+    /// Only one name is allowed
+    if (!res.empty())
+        return;
```

What will happen if I specify multiple names?

---

For duplicates, the behavior is identical to duplicating `path`, if they are the same then the result will be correct, if they are different then the list is empty. This is because after receiving data from the server (no matter which one) we will still apply filtering.

This condition does not affect the result and is simply necessary to avoid doing anything unnecessary.

The `OR` expression does not work in any case, so it does not affect the new feature.

For example:
```sql
# correct result
SELECT * FROM system.zookeeper WHERE path = '/' AND zookeeperName = 'aux_zookeeper' AND zookeeperName = 'aux_zookeeper'

# empty
SELECT * FROM system.zookeeper WHERE path = '/' AND zookeeperName = 'aux_zookeeper' AND zookeeperName = 'default'
```

---

Should we throw an exception if multiple names are specified?

---

For path we also support a kind of recursive/full scan reads (see `allow_unrestricted_reads_from_keeper`) and therefore `OR` conditions for `path` (although it's suboptimal). The user may expect that a query like `select * from system.zookeeper where path='/smth' and (zookeeperName = 'aux_zookeeper' or zookeeperName = 'default')` will read the path from both zookeepers, and it's not something very strange. Not supporting this is okay, but we need a clear exception message instead of unexpected query result that contains data from one zk only

---

If I understand the code correctly, `allow_unrestricted_reads_from_keeper` allows writing LIKE-like masks for paths, but not `path = 'a' or path = 'b'`.

In fact, queries using OR will not work, we just ignore `OR` function_name inside `extractNameImpl` and `extractPathImpl`

---

But you've highlighted an interesting issue. Since we never handle `OR`, your example will behave strangely. In the case of path, code like `where path='/smth' OR path='/smth2'` will throw an exception because it won't find any path.

In my case, it won't find anything either, but it will set the default value, which is clearly confusing.

---

Are you sure that `where path='/smth' OR path='/smth2'` will not work/will throw an exception? Have you tested it with `allow_unrestricted_reads_from_keeper=1`?

---

Oh, you're right. In the case of `allow_unrestricted_reads_from_keeper` it behaves interestingly - `extractPathImpl` will find nothing, set `res.insert({"/", ZkPathType::Recurse});` by default, load all paths recursively and simply filter the results using standard means.

---

I could add two exceptions:
- if the user tries to specify different `zookeerName` multiple times
- if we found `zookeeperName` inside an `or` expression

---

Sorry for the delay. Added an exception as you suggested.